File: BTFViewer/btf_viewer_pkg/disabled_reason.py
"""Disabled-state prerequisite checks (Step 3).

Lockstep with ``web/src/utils/disabledReason.js``.
"""
from __future__ import annotations

from typing import Any, Dict, Optional, Tuple

DISABLED_REASONS: Dict[str, str] = {
    "no_trace": "Open a trace first",
    "two_traces": "Open at least two traces",
    "cursors2": "Place at least two cursors (C1–Cn)",
    "ai_config": "Configure an AI provider in Settings",
    "smp_only": "Requires a multi-core trace",
    "no_evidence": "No Timeline Evidence for this finding",
    "unavailable": "Unavailable",
}


def check_prerequisite(
    requires: str,
    ctx: Dict[str, Any],
    fallback: str = "",
) -> Tuple[bool, str]:
    req = str(requires or "none")
    if not req or req == "none":
        return True, ""
    if req == "trace":
        if ctx.get("has_trace"):
            return True, ""
        return False, fallback or DISABLED_REASONS["no_trace"]
    if req == "two_traces":
        if int(ctx.get("trace_count") or 0) >= 2:
            return True, ""
        return False, fallback or DISABLED_REASONS["two_traces"]
    if req == "cursors2":
        if int(ctx.get("cursor_count") or 0) >= 2:
            return True, ""
        return False, fallback or DISABLED_REASONS["cursors2"]
    if req == "ai_config":
        if ctx.get("ai_configured"):
            return True, ""
        return False, fallback or DISABLED_REASONS["ai_config"]
    if req == "smp":
        if ctx.get("is_multi_core"):
            return True, ""
        return False, fallback or DISABLED_REASONS["smp_only"]
    return True, ""


def build_prerequisite_context(
    *,
    trace: Any = None,
    compare_tab_count: int = 0,
    cursor_count: int = 0,
    ai_configured: bool = True,
) -> Dict[str, Any]:
    cores = 0
    if trace is not None:
        meta = getattr(trace, "meta", None) or {}
        if isinstance(meta, dict):
            cores = len(meta.get("cores") or [])
        elif hasattr(trace, "cores"):
            cores = len(getattr(trace, "cores") or [])
    return {
        "has_trace": trace is not None,
        "trace_count": compare_tab_count,
        "cursor_count": cursor_count,
        "is_multi_core": cores > 1,
        "ai_configured": ai_configured,
    }

File: BTFViewer/web/src/utils/disabledReason.js
/**
 * Disabled-state prerequisite checks (Step 3).
 * Lockstep with btf_viewer_pkg/disabled_reason.py.
 *
 * Aligns toolbar, menus, and command-palette unavailable messaging.
 */

export const DISABLED_REASONS = Object.freeze({
  noTrace: 'Open a trace first',
  twoTraces: 'Open at least two traces',
  cursors2: 'Place at least two cursors (C1–Cn)',
  aiConfig: 'Configure an AI provider in Settings',
  smpOnly: 'Requires a multi-core trace',
  noEvidence: 'No Timeline Evidence for this finding',
  unavailable: 'Unavailable',
})

/**
 * @typedef {Object} PrerequisiteContext
 * @property {boolean} [hasTrace]
 * @property {number} [traceCount]
 * @property {number} [cursorCount]
 * @property {boolean} [isMultiCore]
 * @property {boolean} [aiConfigured]
 */

/**
 * Evaluate a prerequisite token (matches COMMAND_PALETTE_META.requires).
 * @param {string} requires
 * @param {PrerequisiteContext} ctx
 * @param {string} [fallback]
 * @returns {[boolean, string]}
 */
export function checkPrerequisite(requires, ctx, fallback = '') {
  const req = String(requires || 'none')
  if (!req || req === 'none') return [true, '']
  if (req === 'trace') {
    return ctx.hasTrace
      ? [true, '']
      : [false, fallback || DISABLED_REASONS.noTrace]
  }
  if (req === 'two_traces') {
    return (ctx.traceCount ?? 0) >= 2
      ? [true, '']
      : [false, fallback || DISABLED_REASONS.twoTraces]
  }
  if (req === 'cursors2') {
    return (ctx.cursorCount ?? 0) >= 2
      ? [true, '']
      : [false, fallback || DISABLED_REASONS.cursors2]
  }
  if (req === 'ai_config') {
    return ctx.aiConfigured
      ? [true, '']
      : [false, fallback || DISABLED_REASONS.aiConfig]
  }
  if (req === 'smp') {
    return ctx.isMultiCore
      ? [true, '']
      : [false, fallback || DISABLED_REASONS.smpOnly]
  }
  return [true, '']
}

/** Build a prerequisite context from common app refs. */
export function buildPrerequisiteContext({
  trace = null,
  compareTabCount = 0,
  cursorCount = 0,
  aiConfigured = true,
} = {}) {
  const cores = trace?.meta?.cores?.length ?? trace?.cores?.length ?? 0
  return {
    hasTrace: !!trace,
    traceCount: compareTabCount,
    cursorCount,
    isMultiCore: cores > 1,
    aiConfigured,
  }
}
